From d68b018f2d30df2824335b579ee3bff4c10cb1e0 Mon Sep 17 00:00:00 2001
From: Shawn Ferry <sferry@gmail.com>
Date: Tue, 31 Jan 2017 13:26:32 -0500
Subject: [PATCH 1/5] (no-upstream) Solaris auditing support

Patch is not upstream
(maint) Add Solaris auditing support to agent execution

--- puppet-7.21.0/lib/puppet/agent.rb
+++ puppet-7.21.0/lib/puppet/agent.rb
@@ -4,6 +4,10 @@ require_relative '../puppet/util/at_fork
 
 require 'timeout'
 
+if Facter.value(:osfamily) == "Solaris"
+  require 'PuppetAudit'
+end
+
 # A general class for triggering a run of another
 # class.
 class Puppet::Agent
@@ -37,6 +41,12 @@ class Puppet::Agent
 
   # Perform a run with our client.
   def run(client_options = {})
+    # Always start an audit record
+    if Facter.value(:osfamily) == "Solaris"
+        audit_handle = PuppetAudit.new
+        audit_handle.audit_start
+    end
+
     if disabled?
       log_disabled_message
       return
@@ -117,6 +127,12 @@ class Puppet::Agent
     end
     Puppet.notice _("Shutdown/restart in progress (%{status}); skipping run") % { status: Puppet::Application.run_status.inspect } unless block_run
     result
+  ensure
+    # Always stop an audit record
+    if Facter.value(:osfamily) == "Solaris"
+        audit_handle.audit_stop
+        audit_handle.audit_teardown
+    end
   end
 
   def stopping?
--- puppet-7.21.0/lib/puppet/application/apply.rb
+++ puppet-7.21.0/lib/puppet/application/apply.rb
@@ -1,6 +1,9 @@
 require_relative '../../puppet/application'
 require_relative '../../puppet/configurer'
 require_relative '../../puppet/util/profiler/aggregate'
+if Facter.value(:osfamily) == "Solaris"
+  require 'PuppetAudit'
+end
 
 class Puppet::Application::Apply < Puppet::Application
   require_relative '../../puppet/util/splayer'
@@ -176,12 +179,22 @@ Copyright (c) 2011 Puppet Inc., LLC Lice
   end
 
   def run_command
+    # configure auditing on Solaris
+    if Facter.value(:osfamily) == "Solaris"
+       audit_handle = PuppetAudit.new
+       audit_handle.audit_start
+    end
     if options[:catalog]
       apply
     else
       main
     end
   ensure
+    # clean up auditing records on Solaris
+    if Facter.value(:osfamily) == "Solaris"
+      audit_handle.audit_stop
+      audit_handle.audit_teardown
+    end
     if @profiler
       Puppet::Util::Profiler.remove_profiler(@profiler)
       @profiler.shutdown
